Add support for setting sysctls

rhatdan opened the pull request:
This patch will allow users to specify namespace specific "kernel parameters"
for running inside of a container.

Dan Walsh dwalsh@redhat.com

Signed-off-by: Dan Walsh dwalsh@redhat.com

LK4D4 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -370,3 +370,38 @@ func (s *DockerSuite) TestRunSwapLessThanMemoryLimit(c *check.C) {
+// "1" should be printed
+func (s *DockerSuite) TestRunSysctls(c *check.C) {
+
+	out, _, err := dockerCmdWithError("run", "--sysctl", "net.ipv4/ip_forward=1", "--name", "test", "busybox", "cat", "/proc/sys/net/ipv4/ip_forward")
```
Two things. This test doesn't compile, and seems it should be `net.ipv4.ip_forward=1`

LK4D4 commented:
design ok for me
ping @docker/core-maintainers

calavera commented on daemon/config_unix.go:
```diff
@@ -28,6 +28,7 @@ type Config struct {
 	EnableCors           bool
 	EnableSelinuxSupport bool
 	SocketGroup          string
+	Sysctls              map[string]string
```
This field in the daemon configuration is not necessary, is it?

rhatdan replied:
No removed.  Patch built off of Ulimits patch and don't think we need options for the daemon.  Left this by mistake.

rhatdan commented:
Tests should be fixed.

rhatdan commented:
net.ipv4.ip_forward=1 makes more sense but the other form worked also.

LK4D4 commented:
@rhatdan Yup, you're right. It's bug though.

cpuguy83 commented on runconfig/hostconfig.go:
```diff
@@ -249,6 +249,7 @@ type HostConfig struct {
 	RestartPolicy     RestartPolicy         // Restart policy to be used for the container
 	SecurityOpt       []string              // List of string values to customize labels for MLS systems, such as SELinux.
 	ReadonlyRootfs    bool                  // Is the container root filesystem in read-only
+	Sysctls           []string              // List of system controls to be set in the container
```
This should be strongly typed.

cpuguy83 replied:
This I would still prefer to see at least as a map[string]string rather than splitting on `=` to hand off to the exec driver.
Thoughts from anyone else?

rhatdan replied:
I have no problem changing it to this, if others agree.

mrunalp replied:
+1 to making this a map.

cpuguy83 commented on daemon/execdriver/driver.go:
```diff
@@ -204,6 +204,7 @@ type Command struct {
 	Pid                *Pid              `json:"pid"`
 	UTS                *UTS              `json:"uts"`
 	Resources          *Resources        `json:"resources"`
+	Sysctls            map[string]string `json:"sysctls"`
```
This should be strongly typed.

rhatdan replied:
Not sure it can be.  You mean type the value?  I believe this matches up with what runc/libcontainer expects.

Is the value on a sysctl always a number?

cpuguy83 commented:
Is there a list somewhere of namespaced sysctl's?

mrunalp commented:
@cpuguy83 See https://github.com/opencontainers/runc/pull/303

cpuguy83 commented:
@mrunalp Cool, thanks!
Ideally docker should throw an error on an invalid opt before it tries to start the container.

mrunalp commented:
@cpuguy83 Yeah, we could expose a list of valid sysctls via a utility function that could be used both here and in runc/libcontainer.

LK4D4 commented:
@mrunalp Also, I think that sysctls with `/` shouldn't be accepted :)

rhatdan commented:
Once we have a libcontainer function for docker to call to verify the validity of the syscall, I will add it to this code.   One problem with the validation, is that the list is likely to change depending on the kernel you are using.  We could end up blocking valid sysctl's, and eventually allowing some that will not work on the kernel.  Most cases the kernel will probably reject the sysctl.

rhatdan commented:
@jeremyeder do syscalls take anything other then integers?  Do they ever take strings?

rhatdan commented:
@cpuguy83 Seems like sysctl can pass strings.  So not sure we can strongly type the values.  But we can check if the syscall is supported.

```
man sysctl
...
       variable=value
              To  set a key, use the form variable=value where variable is the
              key and value is the value to set it to.  If the value  contains
              quotes or characters which are parsed by the shell, you may need
              to enclose the value in double quotes.   This  requires  the  -w
              parameter to use.
```

jeremyeder commented:
oh yes they take strings.  for example modify kernel.core_pattern in RHEL.

mrunalp commented:
@cpuguy83 IMO, having a string to string hash makes the most sense. If there is an invalid value, then we get back invalid argument error trying to write it under /proc/sys

cpuguy83 commented:
This makes sense.

rhatdan commented:
@cpuguy83 @mrunalp @LK4D4 PTAL

Now just passes around the map[string]string.

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -374,3 +374,38 @@ func (s *DockerSuite) TestRunSwapLessThanMemoryLimit(c *check.C) {
+// "1" should be printed
+func (s *DockerSuite) TestRunSysctls(c *check.C) {
+
+	out, _, err := dockerCmdWithError("run", "--sysctl", "net.ipv4.ip_forward=1", "--name", "test", "busybox", "cat", "/proc/sys/net/ipv4/ip_forward")
```
Can we use `dockrCmd` here?

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -374,3 +374,38 @@ func (s *DockerSuite) TestRunSwapLessThanMemoryLimit(c *check.C) {
+		c.Fatalf("setting the sycstls net.ipv4.ip_forward failed got %q", out)
+	}
+
+	out, _, err = dockerCmdWithError("run", "--sysctl", "net.ipv4.ip_forward=0", "--name", "test1", "busybox", "cat", "/proc/sys/net/ipv4/ip_forward")
```
`dockerCmd`

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -374,3 +374,38 @@ func (s *DockerSuite) TestRunSwapLessThanMemoryLimit(c *check.C) {
+	out, err = inspectFieldJSON("test", "HostConfig.Sysctls")
+	c.Assert(err, check.IsNil)
+
+	if out != "[\"net.ipv4.ip_forward=1\"]" {
```
I think this test will fail now with the map?

cpuguy83 replied:
Also, `c.Assert`

cpuguy83 commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -196,6 +196,7 @@ Create a container
              "RestartPolicy": { "Name": "", "MaximumRetryCount": 0 },
              "NetworkMode": "bridge",
              "Devices": [],
+             "Sysctls": [{}],
```
This needs to be updated to use the map.

cpuguy83 commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -297,6 +298,7 @@ Json Parameters:
     -   **Devices** - A list of devices to add to the container specified as a JSON object in the
       form
           `{ "PathOnHost": "/dev/deviceName", "PathInContainer": "/dev/deviceName", "CgroupPermissions": "mrw"}`
+    -   **Sysctls** - A list of kernel parameters (sysctls) to set in the container, in the form of `VAR=value`
```
Also this

cpuguy83 commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -417,6 +419,7 @@ Return low-level information on the container `id`
 				"Type": "json-file"
 			},
 			"SecurityOpt": null,
+			"Sysctls": [{}],
```
Also this

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -374,3 +374,38 @@ func (s *DockerSuite) TestRunSwapLessThanMemoryLimit(c *check.C) {
+		c.Fatalf("failed to run container: %v, output: %q", err, out)
+	}
+	out = strings.TrimSpace(out)
+	if out != "1" {
```
`c.Assert(strings.Trimspace(out), check.Equals, "1")`

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -374,3 +374,38 @@ func (s *DockerSuite) TestRunSwapLessThanMemoryLimit(c *check.C) {
+		c.Fatalf("failed to run container: %v, output: %q", err, out)
+	}
+	out = strings.TrimSpace(out)
+	if out != "0" {
```
c.Assert

cpuguy83 commented on integration-cli/docker_cli_run_unix_test.go:
```diff
@@ -374,3 +374,38 @@ func (s *DockerSuite) TestRunSwapLessThanMemoryLimit(c *check.C) {
+
+// "1" should be printed
+func (s *DockerSuite) TestRunSysctls(c *check.C) {
+
```
`testRequires(c, NativeExecDriver)`

rhatdan commented:
@cpuguy83 I think I fixed all issues.

cpuguy83 commented:
LGTM

For the API docs, probably also need to add a note that the values are specifically string values only.

vdemeester commented:
LGTM

cpuguy83 commented:
Moving to docs review, ping @thaJeztah

thaJeztah commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -297,6 +298,10 @@ Json Parameters:
       form
           `{ "PathOnHost": "/dev/deviceName", "PathInContainer": "/dev/deviceName", "CgroupPermissions": "mrw"}`
+    -   **Sysctls** - A list of kernel parameters (sysctls) to set in the container, specified as
+          `{ "<name>: <Value> }`, for example:
```
Missing some quotes here, or remove the quote

thaJeztah commented on docs/reference/api/docker_remote_api_v1.21.md:
```diff
@@ -417,6 +422,9 @@ Return low-level information on the container `id`
 			"SecurityOpt": null,
+			"Sysctls": {
+			        "net.ipv4.ip_forward": "1"
+                        },
```
just a nit; does this curly bracket align properly? Could be GitHub isn't rendering it properly

thaJeztah commented:
Thanks @rhatdan I think we need docs in some more places;

- A mention in the API changes; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md
- The docker create CLI docs; https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md
- The docker run CLI docs; https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md

Possibly, the docker run reference; https://github.com/docker/docker/blob/master/docs/reference/run.md

rhatdan commented:
Fixed all except     docs/reference/run.md

I changed everywhere ulimits was mentioned and it is not mentioned here.  Perhaps we should add both ulimits and sysctl here?

cpuguy83 commented:
@rhatdan Do you agree this closes #4717?

rhatdan commented:
Yes I believe that this and the ulimits patches solve the problems discussed there.

jfrazelle commented:
needs rebase :/

thaJeztah commented on docs/reference/api/docker_remote_api.md:
```diff
@@ -76,6 +76,7 @@ This section lists each version from latest to oldest.  Each listing includes a
 
 [Docker Remote API v1.21](/reference/api/docker_remote_api_v1.21/) documentation
 
+* `POST /containers/create` and `POST /containers/(id)/start`allow you toconfigure kernel parameters (sysctls) for use in the container.
```
nits:
- missing space before "allow"
- s/toconfigure/to configure/

thaJeztah commented on docs/reference/commandline/run.md:
```diff
@@ -540,3 +541,9 @@ the three processes quota set for the `daemon` user.
+
+The `--sysctl` flag sets the namespaces kernel parameters (sysctls) in the container, specified as
+          `{ <name>: <Value> }`, for example:
+	  `{ "net.ipv4.ip_forward": "1" }`
```
Re-reading this, can you make this a "full" example, as I'm currently in doubt if the `{}` should be used on the command line, or not

i.e. currently it reads that it should be used as:

    docker run --sysctl '{ "net.ipv4.ip_forward: "1" }' someimage

Is that correct? Or should it be:

    docker run --sysctl net.ipv4.ip_forward=1 someimage

rhatdan commented:
@thaJeztah I fixed the docs, and also added a little more information no run.md about checking whether a kernel parameter/sysctl is namespaced.

thaJeztah commented on docs/reference/commandline/run.md:
```diff
@@ -541,3 +542,14 @@ the three processes quota set for the `daemon` user.
+
+Note: Not all sysctls are namespaced.  You should make sure that modifying a
+sysctl inside of a container does not modify it on the host system.  As the
+kernel evolves more sysctls will become namespaced.
```
I saw some formatting/Markdown issues here, could you use the text below?

```Markdown
The `--sysctl` sets namespaced kernel parameters (sysctls) in the
container. For example, to turn on IP forwarding in the containers
network namespace, run this command:

    $ docker run --sysctl net.ipv4.ip_forward=1 someimage


> **Note**: Not all sysctls are namespaced. You should make sure that
> modifying a sysctl inside of a container does not modify it on the host
> system. As the kernel evolves more sysctls will become namespaced.
```

thaJeztah commented:
Thanks, @rhatdan the extra information is useful. I noticed some formatting issues, see my comment above

thaJeztah commented:
docs LGTM, thanks @rhatdan

moxiegirl commented:
LGTM thank you @rhatdan

vdemeester commented:
@thaJeztah mergeable, needs some impact labels right ?

thaJeztah commented:
@vdemeester do not merge yet; we want the PR's that are on the milestone first (in case there will be merge conflicts)

thaJeztah commented:
Temporarily moved back to code review; there were some concerns about this feature and more eyes needed to review this after code freeze

rhatdan commented:
Ok, what were the concerns?

crosbymichael commented:
@rhatdan Isn't there security issues with providing an API to set sysctls because they are not all namedspaced and affect settings on the host? 

To the normal user they will think these only apply to the container but in reality the majority of these sysctls will be applied to the host and call containers running on it.

rhatdan commented:
@crosbymichael Sure, but we don't provide security at the docker run level, just within the container.  
Users can get them selves into problems by executing --priviliged or -v /dev:/dev or ...

My problem with providing validation, is that it is a moving target, and as soon as it moves, we will have some kernels that use global settings, while the same sysctl on a different kernel will be namespaced.  One way or another we will be wrong, either preventing or allowing.  I think this should be up to user to understand which is which.

crosbymichael commented:
Could you give some of the use cases for this?  I would like to know what type of settings people are expecting to change for the container.

rhatdan commented:
@jeremyeder do you have use cases?

rhatdan commented:
Just googling docker sysctl site:bugzilla.redhat.com
Finds a couple 
sysctl -w kernel.shmmax=1000000000
net.ipv6.conf.all.forwarding = 1

rhatdan commented:
https://github.com/docker/docker/issues/4717 has other examples.

cpuguy83 commented:
See #4717

rhatdan commented:
We have another customer requesting
All IPC related settings:

-        msgmni
-        msgmnb
-        msgmax
-        shmmni
-        shmmax
-        shmall

File descriptor settings, both via ulimit settings and fs.file-max parameter.

Network parameters as well, at least:

net.ipv4.tcp_rmem
net.ipv4.tcp_wmem

pmoust commented:
We also need shmmax / shmall. 
Needs rebase

rhatdan commented:
@crosbymichael Any word on this?

rhatdan committed:
Add support for setting sysctls

This patch will allow users to specify namespace specific "kernel parameters"
for running inside of a container.

Dan Walsh <dwalsh@redhat.com>

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

crosbymichael commented:
I'm -1 on this.  There is currently too much room for user error to mess things up.  If you know what settings to make on these then you can exec into your container and set them yourself.

cpuguy83 commented:
That's true, now that we have `--privileged` on exec especially.

mrunalp commented:
Using exec is racy though as the container process could fail before the exec is even called.

rhatdan commented:
So is there no road forward on this?  There is potential race conditions, where you could not get in, on time to set the sysctl, if pid 1 needed the access.

rhatdan commented:
Seems to lots of ways for users to screw up their system, if they do not know what they are doing.  Not sure why I would block just this way.

calavera commented:
I think there is a general agreement that this is potentially dangerous for users. It looks like other maintainers also agree with that feeling. Sorry, but I'm going to close this PR and we won't move it forward.

rhatdan commented:
We are hearing a lot from customers about the need for this. If we add validation would the pull request be reconsidered?  @calavera @crosbymichael 

Reopen https://github.com/opencontainers/runc/pull/303 for runc and then add validation on the cli side also.

mrunalp commented:
I think https://github.com/opencontainers/runc/pull/303 is a good list and should open up setting many sysctls that are useful in a container.

gordillo-ramon commented:
Hi. This is one customer that thinks that this is an absolutely needed feature. Otherwise, some technologies like redis are at risk of DoS when running in a container.

jfrazelle commented:
I still believe this is going to be a cause for user confusion when they realize sysctls are not namespaced

mrunalp commented:
@jfrazelle We are proposing that there be a whitelist like in the runc PR and then we can return a nice error message to the user when they try to set something that isn't namespaces. A lot of container applications need to tune the network sysctls (which are namespaced) and not allowing this feature is an adherent to containerizing such applications.

jfrazelle commented:
ok that seems more reasonable :)

On Mon, Jan 11, 2016 at 4:45 PM, Mrunal Patel <notifications@github.com>
wrote:

> @jfrazelle <https://github.com/jfrazelle> We are proposing that there be
> a whitelist like in the runc PR and then we can return a nice error message
> to the user when they try to set something that isn't namespaces. A lot of
> container applications need to tune the network sysctls (which are
> namespaced) and not allowing this feature is an adherent to containerizing
> such applications.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16632#issuecomment-170744431>.
>

rhatdan commented:
Right we can document the sysctls that are currently available and then expand the list as the kernel evolves.  Can we get this pull request reopened, or should I open a new one?

thaJeztah commented:
@rhatdan looks like the branch was updated/force-pushed while it was closed, so GitHub doesn't allow me to reopen the PR

rhatdan commented:
Ok I will open a new pull request

mnp commented:
Is there an issue to support the `fs.*` tree as well?  At least obvious things like fs.file-max and fs.nr_open? It doesn't look like those are in the https://github.com/docker/docker/pull/16632/commits/7dff8cce0779d0fb78be692bb7430a73efdc8bc7 merge.   Thanks.

cpuguy83 commented:
@mnp If you test these out and find that they are namespaced (which I doubt they are, but maybe), then you can create an issue.
If they are not namespaced, we cannot support them.

mnp commented:
@cpuguy83 Yes, by testing only, it does not seem that fs.* are namespaced in our 3.10 centos kernel at least.  Containers inherit the host settings.  Thanks.
